### ISSUE TITLE
Consider using `bytemuck` as replacement for `read_fonts::FromBytes`

### ISSUE THREAD

**Issue by drott** [googlefonts/fontations#808]
@anforowicz asks

> Can we remove/delete `read_fonts::FromBytes` and replace it with `bytemuck::AnyBitPattern`?

Reasoning:

> It seems to me that the [`read_fonts::FromBytes`](https://docs.rs/read-fonts/latest/read_fonts/trait.FromBytes.html) unsafe trait is a reimplementation of the [`bytemuck::AnyBitPattern`](https://docs.rs/bytemuck/latest/bytemuck/trait.AnyBitPattern.html) trait (see also [`bytemuck::from_bytes`](https://docs.rs/bytemuck/latest/bytemuck/fn.from_bytes.html) function).

Motivation / Benefits:

- Code reuse: `bytemuck` already provides auto trait impls and casts for various things (slices of T, mut slices of T, values of T, etc).  No need to reimplement this.
- Audit reuse: `bytemuck`'s `unsafe` usage has already been reviewed (e.g. see [go/image-crate-chromium-security-review](https://goto.google.com/image-crate-chromium-security-review)).  Removing some (most? all?) uses of `unsafe` in the `read_fonts` help will help to streamline future audits of this crate.

**Comment by drott** [comment]
@anforowicz - on the second point, could you explain a bit what you include and foresee by "future audits", do you mean Google audits or audits by external parties as well?

For context, this was the original security review of `read_fonts` and fontations in general: [Internal doc](https://docs.google.com/document/d/1Whi4ONnDPdeBqXUViRLMaLTtLkFE_Qs2xRXZl0vDfYo/edit)

Additional context, unsafe / soundness discussion on `read_ref_unchecked` https://github.com/googlefonts/fontations/issues/436

I'll let Fontations folks comment, but my 2 cents:

- `read_fonts::FromBytes` is somewhat foundational code that we expect no churn or much any changes on in the future, so the likeliness of this having to be reviewed again is low. In my understanding means that we can build on the previous security review chain and do not have to revisit this, would you agree?
- We've so far been intentionally conservative with adding dependencies in order to ease adoption of the library in other open-source contexts. `byte_muck` does seem to be widely available and used, but I am not sure adding a dependency to this library for this single use-case might not be the right-tradeoff? 
  - If I am reading it right, bytemuck has [not been vetted as an external dependency for Chromium yet](https://source.chromium.org/chromium/chromium/src/+/main:third_party/rust/chromium_crates_io/vendor/).
  - IIUC correctly from a separate discussion in chat @dfrg  mentions, there are slight differences between `FromBytes` and `AnyBitPattern`, and this was looked at during #436 and the initial security review - @dfrg can you elaborate, do you have any other references? @cmyr mentions that our raw read use case is more constrained and is not expected to extend in scope.

**Comment by cmyr** [comment]
I believe it would be relatively straight-forward to replace our internal trait with bytemuck. The main downside is that this involves bringing in a substantial dependency; this is not the end of the world.

I think that using bytemuck would also remove one possible avenue for unsoundness, which would be a bug in our codegen implementation which incorrectly marks a type as `FromBytes`; if memory serves bytemuck uses a derive macro to implement this trait, and that implementation involves some const trickery to verify the relevant invariants (alignment, lack of padding.) Our own codegen implements this trait directly where it determines it is appropriate. I think it would be quite unlikely given the nature of the types in question, but it would not be impossible for us to introduce a bug that implemented this trait incorrectly (although it would require an additional step for us to actually trigger UB by trying to cast bytes to this mis-labeled type, which also only happens in generated code.)

So... I think that from a safety perspective I am satisfied with the status quo, but I do on balance there's a good argument for adopting bytemuck, and I'm happy to write a patch to try it out, and at least see that we don't run into any unexpected complications.

**Comment by rsheeter** [comment]
> It would be great to encapsulate the unsafe code into a crate that (almost) never changes

This idea appeals to me.

**Comment by cmyr** [comment]
Mentioned in chat, but I don't know that the 'encapsulate unsafe code in separate crate' idea is really that useful here. Regardless of where it is defined, the unsafe trait will be implemented in codegen, and that is the place where a series of bugs could introduce unsoundness; and there are additional annoyances with having a separate crate, such as having to deal with the orphan rules (a nice thing about us defining this trait in read-fonts is that we can implement it for types in font-types)

**Comment by rsheeter** [comment]
Update from chat: If bytemuck gets rid of *all* our unsafe that's appealing. I initially thought it would get rid of only _some_.

**Comment by drott** [comment]
Requested `bytemuck` inclusion to Chromium in https://issues.chromium.org/328949638 - if possible let's hold landing a change that makes bytemuck mandatory, no objections on optional dep) before we have an ETA on bytemuck inclusion to Chromium.

**Comment by drott** [comment]
WIP CL on `bytemuck` inclusion into Chromium https://chromium-review.googlesource.com/c/chromium/src/+/5362675 - thanks @anforowicz

**Comment by drott** [comment]
Did we conclude the discussion on whether this should be an optional or mandatory dependency? Now that this landed in Chromium and received an unsafe review, from Chromium's side, I see no obstacles to making this mandatory.

**Comment by cmyr** [comment]
yes, our intention is to use this to remove all unsafe from `read-fonts` and `skrifa`.

**Comment by drott** [comment]
So, non-optional?

**Comment by drott** [comment]
On the Chrome Rust chat there was a brief discussion on the `zerocopy` crate vs. the `bytemuck` crate and maintainership of the crates being a bit more on the Google side for `zerocopy` side, so I wondered whether a transition to zerocopy is needed. After discussing with @anforowicz we believe no action is necessary. Copying notes of @anforowicz assessment below:

* If Rust code png work resumes on Chromium that depends on bytemuck - this is kind of an argument that `bytemuck` is not going anywhere.

* AFAIU Fontations mostly uses `bytemuck::AnyBitPattern` and `bytemuck::NoUninit` which then help with the usage of `bytemuck::try_cast_slice_mut`.  So if needed, transitioning to `zerocopy` shouldn't require much work - you could just switch to `zerocopy::FromBytes` and https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html#method.mut_slice_from AFAIU.  https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html#method.mut_slice_from_prefix is a nice convenience method that would let you simplify your code a tiny bit.

I think the security properties of bytemuck and zerocopy are both good enough.  And performance seems the same in your case.

**Comment by dfrg** [comment]
Looking at `zerocopy` again, there might be a potential win here in that it seems like we could remove all of our unsafe code since the `FromBytes` derive macro is less sensitive to the things that trip up `bytemuck`. I think this is worth exploring.
